subctl/diagnore-pre-install.md

```diff
@@ -0,0 +1,102 @@
+# Subctl `diagnose pre-install`
+
+## Summary
+
+As a user of Submariner, it would be useful in a myriad of situations to have a tool capable of testing the Kubernetes infrastructure
+where Submariner is to be installed.
+The tool should check all of Submariner's pre-requisites from the infrastructure, and report any deficiencies.
+It would be nice if the tool can also offer guidance for the deficiencies, so that the user can address them and install Submariner successfully.
+
+This tool would compliment other `subctl diagnose` commands which check that Submariner is installed and working properly, by allowing
+similar checks **before** the actual installation.
+
+Alternatively, the new command could be hosted elsewhere, if nesting under `diagnose` is deemed confusing.
+
+## Proposal
+
+Add a `subctl diagnose pre-install` command which checks that the infrastructure supports installing Submariner by checking:
+
+* Gateway nodes are set-up.
+  * Kube proxy, if there, doesn't run in IPVS mode - as it's currently not supported.
+* The broker is reachable from the cluster.
+* K8s version is supported.
+* The CNI is supported.
+* Conflicts with existing clusters in the clusterset.
+  * CIDRs should be distinct, unless the clusterset is using globalnet.
```

Comment:
Perhaps I am missing something, but at the point where the tool is going to be used there isn't a clusterset...?

Reply:
There could already be a clusterset, if it's not the first cluster you're adding..

In fact, the common case would be that you already have a clusterset, unless it's the first cluster you're ever adding.

Anyway I'll add clarification.

Reply:
The number one use case I was thinking of for this tool is greenfield deployments of Submariner. I can see your point as well.

Reply:
The number one use case I was thinking of for this tool is greenfield deployments of Submariner. I can see your point as well.

Reply:
From a workflow perspective, though, you'd deploy the broker before the clusters, so you'd still have an empty clusterset, and as you may know from set theory, an empty set is still a set ;)